rib/rib.go

```diff
@@ -1543,6 +1543,22 @@ func (r *RIBHolder) doDeleteMPLS(label uint32) {
 	delete(r.r.Afts.LabelEntry, aft.UnionUint32(label))
 }
 
+// locklessDeleteMPLS removes the label forwarding entry with the specified label
+// from the RIB, without holding the lock on the AFT. The calling routine
+// MUST ensure that it holds the lock to ensure thread-safe operation.
+func (r *RIBHolder) locklessDeleteMPLS(label aft.Afts_LabelEntry_Label_Union) error {
+	de := r.r.Afts.LabelEntry[label]
+	if de == nil {
+		return fmt.Errorf("cannot find label %d", label)
+	}
+
+	delete(r.r.Afts.LabelEntry, label)
+	if r.postChangeHook != nil {
+		r.postChangeHook(constants.Delete, unixTS(), r.name, de)
```

Comment:
This is probably ok for now, but given that the rate of delete can get quite high would it be worthwhile to process these hooks async?

Reply:
It definitely is possible to do this -- it might be better to refactor this out from a callback design to just taking a channel that we write to, and doing a non-blocking channel write instead. I'll file and issue to consider this.

Reply:
Added in https://github.com/openconfig/gribigo/issues/217.